### PR TITLE
Feature: urllib3 instead of curl

### PR DESCRIPTION
**Pull request by spawn-guy**
for https://github.com/celery/kombu/issues/737

i've migrated `CurlClient` to `Urllib3Client` and updated some unit tests.

i'd like to test this more thoroughly, though

needs a little bit more code cleanup as i am not even use this works (tests don't fail at least)

let me know what you think

### CODE REVIEW

**Review by Nusnus — CHANGES_REQUESTED** [review]
Hey there 👋 
Can you please the CI failures?

Thanks 🙏

**Review by auvipy — CHANGES_REQUESTED** [review]
this would be a good move. please also check the failing tests

File: requirements/docs.txt
Comment (Nusnus) [inline requirements/docs.txt:6]: I love this change 🙏

**Comment by spawn-guy** [comment]
a problem i see is that `urllib3.PoolManager` doesn't support proxy per call. to support proxies i should instantiate `urllib3.ProxyManager`.

i think i need some guidance on how to properly test this migration: Who uses this client Outside of "kombu in vacuum"?

also the curl files are not removed yet

**Review by Nusnus — COMMENTED** [review]
I gave it a quick review, nothing special popped into my eyes so far.
I'll review again if the CI fully passes.

**Comment by Nusnus** [comment]
> a problem i see is that `urllib3.PoolManager` doesn't support proxy per request. to support proxies i should instantiate `urllib3.ProxyManager`.
> 
> i think i need some guidance on how to properly test this migration: Who uses this client Outside of "kombu in vacuum"?
>

Good question. Maybe check Celery code?

> also the curl files are not removed yet
>

Can you change the PR to draft meanwhile?
Let us know it's done when you set it back to "Ready for Review" 🙏

**Comment by spawn-guy** [comment]
i've run my code inside my `beta` environment (or at least i hope i did.. `pipenv` seems to checkout the correct commit id from my repo) (i even verified that the `urllib3_client.py` is, indeed, in the `site-packages` folder)

i've also tried `pytest-celery` and tests in `celery` (where there is a weird bug on Windows with some `{docstring}` and unicode characters - so i removed the `{docstring}`) and tests succeeded with `Tests passed: 32,167, ignored: 160 of 32,331 tests`

and all seems to behave as usual.

@auvipy @Nusnus the last questions are:
- what do we do with the `proxy-per-request`? a 1 `urllib3.ProxyManager` with 1 proxy per Pool ... or should i create some more pools per each proxy setting in requests (some lru cache, maybe)?
- Should i continue to clean up the rest of the leftover `curl` code?

**Comment by auvipy** [comment]
> * what do we do with the `proxy-per-request`? a 1 `urllib3.ProxyManager` with 1 proxy per Pool ... or should i create some more pools per each proxy setting in requests (some lru cache, maybe)?

may be the first approach would be fine to start. but if you can also go extra mile, it would be great as well

**Comment by spawn-guy** [comment]
Speaking about refactoring : if I remove the curl=None from request call - we can have a proper proxy pool. I just need to figure out which configure settings are used

**Comment by Nusnus** [comment]
> Speaking about refactoring : if I remove the curl=None from request call - we can have a proper proxy pool. I just need to figure out which configure settings are used. But this is a breaking change 😬

Apologies for not being so helpful, I am pretty focused on the asyncio effort so I have less resources to support this PR as well at the moment 🙏

**Comment by spawn-guy** [comment]
@Nusnus no worries.

i think i'm done :) 
no "max size of pools", though. but the rest of the missing functionality is back :)

i'm gonna deploy this to my testing env

**Comment by spawn-guy** [comment]
so.. apparently, for `aws.getrequest` - someone disabled certificate verification. 

and now (with all my fancy code that enabled missing `request.parameters`) urllib3 complains that ssl verification is disabled :) 

fixing that based on boto code

**Review by Nusnus — COMMENTED** [review]
Ping me when the CI is green.
I’ll give it at least a quick review to see nothing pops into my eyes.

Thanks!

**Comment by Nusnus** [comment]
Ready for review? @spawn-guy

**Comment by spawn-guy** [comment]
@Nusnus Ready 🤓 

i have this running on my aws environment. and i was debugging the errors on it (and i was, indeed, getting urllib3 errors) - so i know this thing is Working 😝 

but, i don't use any proxy (duh) - this is the part that can have errors in the field applications. and i'm ready to fix them

File: kombu/asynchronous/aws/connection.py
Comment (spawn-guy) [reply kombu/asynchronous/aws/connection.py:96]: the issue is a bit more complex and multi-layered. i am going to lay out my timeline and go a bit deeper into my reasoning.

- i implemented urllib3 Pool with default settings. and totally missed all the extra `Request.*` parameters. and no proxy support
- then i implemented proxy support and the missed `Request` parameters. and then i started to test these changes on my aws environment (not production, so i can break it)

with the first implementation of parameters i started to get these urllib3 warnings:
```
[2024-10-08 15:14:15,172: WARNING/MainProcess] /var/app/venv/staging-LQM1lest/lib64/python3.11/site-packages/urllib3/connectionpool.py:1099: InsecureRequestWarning: Unverified HTTPS request is being made to host 'sqs.eu-west-1.amazonaws.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
```
and i don't like Warnings,.. and they are related to my new code, so i started investigating.

at first i saw the default value on class `Request` `validate_cert = True` in `http/base.py`. so i thought the problem is somewhere else... 

then i saw that in `aws/connection.py` the `validate_cert=` is set to `False`. so i changed this to `True`(i mentioned it here in comments)... 

then i started to get other errors, that complain about `ca_certs`. this issue was my mistake(1 extra bad line of config). but i fixed in a more proper/better way:

As we already have a `botocore` dependency, and it relies on urllib3 - i decided to see how they handle the connection.
they use `get_cert_path` method. which at first tries to use `certifi.where` and if `certifi` is not present - use their own pre-bundled `cacert.pem`.

here in `kombu` we have 2 layers: 1st in `http/base` for general use; and 2nd in `aws/connection` for AWS use on top of base.
then i split the fix into 2 layers:
- if there is no "custom" certificate in the `Request` - `http/base` will first try to find `ceritifi`(which doesn't require `certifi` dependency), and if not present - keep the default `urllib3.ca_certs` value (i.e. not override with anything - use default). and if there is a custom `Request.ca_certs` - it will set it. resulting in: (optional) `certifi` dependency and no `botocore` dependency
- and on `aws/connection`, as we already depend on `botocore` - i decided to import and use `ca_certs=get_cert_path(True)` with the default `botocore` behaviour. which is a bit more advanced than in `http/base`

and only afterwards i found the recent Issues and changes, that Disable certificate validation for aws connection. 
https://github.com/celery/kombu/issues/2113 and https://github.com/celery/kombu/pull/2114
but they were about `pycurl` which we don't heave anymore ;) and i think it is not wise to reduce security level , just to hotfix an issue.

technically, this PR is 2 PRs:
- replace `pycurl` with `urllib3`
- re-enable ssl verification for AWS requests (now on urllib3 as in botocore and also without it)

File: kombu/asynchronous/http/urllib3_client.py
Comment (spawn-guy) [reply kombu/asynchronous/http/urllib3_client.py:30]: i'll look into this tomorrow

File: kombu/asynchronous/aws/connection.py
Comment (Nusnus) [reply kombu/asynchronous/aws/connection.py:96]: Wow. What a great explanation!
Thank you very much, great great work!! 👏

File: kombu/asynchronous/http/urllib3_client.py
Comment (Nusnus) [reply kombu/asynchronous/http/urllib3_client.py:30]: Ping me when you get it fixed, but it’s not super critical.
I am going to release a new Kombu version (RC) and I’d love to have this new change included.

I plan to release it tomorrow, so if you can get this done tomorrow that would be great.
If not, I’ll merge it anyways, and you can close the test gap in a different PR.

I don’t want to miss the chance to include it so Kombu v5.5 will be awesome!

**Review by Nusnus — APPROVED** [review]
Amazing work!
Thank you very much!

I left a comment or two, check it out before we merge the PR.

**Comment by Nusnus** [comment]
@auvipy I plan to merge it tomorrow and then release a new pre-release for Kombu.
You have requested the CI will pass, it now passes.

Please let us know if there’s anything else required for your approval.
Thank you.

File: kombu/asynchronous/http/urllib3_client.py
Comment (spawn-guy) [reply kombu/asynchronous/http/urllib3_client.py:30]: @Nusnus i've added a few tests. should cover this `_get_pool_key_parts` method

**Pull request merged**
